File: bench/kernels/d1_utf8_validate/ref.c
/* d1-utf8-validate (family D), NAIVE C: the same structural UTF-8 scan,
 * byte at a time, over the same buffer.
 * Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static const char CHUNK[] =
    "wolf pack ranges wide: \xc3\xa9 and \xe2\x82\xac and \xf0\x9f\x90\xba"
    " keep the scan honest. ";
#define REPS 15000

static int64_t validate(const unsigned char *p, size_t len) {
    int64_t want = 0, chars = 0;
    for (size_t i = 0; i < len; i++) {
        unsigned char b = p[i];
        if (want > 0) {
            if (b < 128 || b > 191) return -1;
            want--;
        } else if (b < 128) {
            chars++;
        } else if (b < 194) {
            return -1;
        } else if (b < 224) {
            want = 1; chars++;
        } else if (b < 240) {
            want = 2; chars++;
        } else if (b < 245) {
            want = 3; chars++;
        } else {
            return -1;
        }
    }
    return want == 0 ? chars : -1;
}

int main(int argc, char **argv) {
    int64_t ops = argc > 1 ? (int64_t)strtoull(argv[1], 0, 10) : 20;
    size_t clen = strlen(CHUNK);
    size_t len = clen * REPS;
    unsigned char *buf = malloc(len + 1);
    for (int i = 0; i < REPS; i++) memcpy(buf + (size_t)i * clen, CHUNK, clen);
    buf[len] = 0;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    int64_t sink = 0;
    for (int64_t k = 0; k < ops; k++) sink = validate(buf, len);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    free(buf);
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%llu,\"sink\":%lld}\n",
           (unsigned long long)ns, (unsigned long long)((uint64_t)ops * len),
           (long long)sink);
    return 0;
}

File: bench/kernels/d1_utf8_validate/ref.rs
//! d1-utf8-validate (family D), Rust reference: the same hand-written
//! structural scan (NOT `str::from_utf8`, which is a hand-vectorized
//! library routine — comparing a hand loop against a tuned library would
//! be a benchmarking crime, see bench/protocol.md).
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;
use std::time::Instant;

const CHUNK: &str = "wolf pack ranges wide: é and € and 🐺 keep the scan honest. ";
const REPS: usize = 15000;

fn validate(p: &[u8]) -> i64 {
    let mut want: i64 = 0;
    let mut chars: i64 = 0;
    for &b in p {
        if want > 0 {
            if b < 128 || b > 191 {
                return -1;
            }
            want -= 1;
        } else if b < 128 {
            chars += 1;
        } else if b < 194 {
            return -1;
        } else if b < 224 {
            want = 1;
            chars += 1;
        } else if b < 240 {
            want = 2;
            chars += 1;
        } else if b < 245 {
            want = 3;
            chars += 1;
        } else {
            return -1;
        }
    }
    if want == 0 { chars } else { -1 }
}

fn main() {
    let ops: u64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(20);
    let text = CHUNK.repeat(REPS);
    let bytes = text.as_bytes();
    let t = Instant::now();
    let mut sink: i64 = 0;
    for _ in 0..ops {
        sink = validate(black_box(bytes));
    }
    let ns = t.elapsed().as_nanos();
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink}}}",
        ops * bytes.len() as u64
    );
}
